### PR TITLE
support for 1.9 extensions

### PR DESCRIPTION
**Pull request by longemen3000**
moves `SpecialFunctions`, `Unitful` and `Juno` support into extensions.

### CODE REVIEW

File: ext/MeasurementsJunoExt.jl [hunk 1]
```diff
@@ -0,0 +1,28 @@
+module MeasurementsJunoExt
+
+@static if isdefined(Base, :get_extension)
```
Comment (giordano) [inline ext/MeasurementsJunoExt.jl:21]: Top-level `@static` does nothing, the expression is "statically" evaluated anyway
```suggestion
if isdefined(Base, :get_extension)
```

File: ext/MeasurementsJunoExt.jl [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        i = -i
+        sss = " - "
+    else
+        sss = " + "
+    end
+    Juno.render(ji, Juno.Row("(", Juno.render(ji, r), ")", sss,
+                            "(", Juno.render(ji, i), ")im"))
+end
+
+end
```
Comment (giordano) [inline ext/MeasurementsJunoExt.jl:46]: Missing newline
```suggestion
end

```

File: ext/MeasurementsSpecialFunctionsExt.jl [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+### Commentary:
+#
+# This file contains definition of special functions that support
+# Measurement objects.
+#
+### Code:
+
+### Special functions
+using .SpecialFunctions
+@static if isdefined(Base, :get_extension)
```
Comment (giordano) [inline ext/MeasurementsSpecialFunctionsExt.jl:22]:
```suggestion
if isdefined(Base, :get_extension)
```

File: ext/MeasurementsSpecialFunctionsExt.jl [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+# Deprecations
+function Base.factorial(a::Measurement)
+    Base.depwarn("""
+                 factorial(($(a))::Measurement) is deprecated, instead use
+                     SpecialFunctions.gamma($(a + one(a)))
+                 """,
+                 :factorial)
+    SpecialFunctions.gamma(a + one(a))
+end
+end
```
Comment (giordano) [inline ext/MeasurementsSpecialFunctionsExt.jl:240]:
```suggestion
end

```

File: ext/MeasurementsUnitfulExt.jl [hunk 5]
```diff
@@ -1,3 +1,13 @@
+module MeasurementsUnitfulExt
+
+@static if isdefined(Base, :get_extension)
```
Comment (giordano) [inline ext/MeasurementsUnitfulExt.jl]:
```suggestion
if isdefined(Base, :get_extension)
```

File: src/Measurements.jl [hunk 6]
```diff
@@ -85,9 +83,14 @@ end
 # Start from 1, 0 is reserved to derived quantities
 const tag_counter = Threads.Atomic{UInt64}(1)
 
-function __init__()
-    @require Unitful="1986cc42-f94f-5a68-af5c-568840ba703d" include("unitful.jl")
-    @require SpecialFunctions="276daf66-3868-5448-9aa4-cd146d93841b" include("special-functions.jl")
+@static if !isdefined(Base, :get_extension)
```
Comment (giordano) [inline src/Measurements.jl:86]:
```suggestion
if !isdefined(Base, :get_extension)
```

File: src/show.jl [hunk 7]
```diff
@@ -92,24 +92,4 @@ end
...
-            i = -i
-            sss = " - "
-        else
-            sss = " + "
-        end
-        Juno.render(ji, Juno.Row("(", Juno.render(ji, r), ")", sss,
-                                 "(", Juno.render(ji, i), ")im"))
-    end
-end
+end
```
Comment (giordano) [inline src/show.jl:95]:
```suggestion
end

```

File: test/runtests.jl [hunk 8]
```diff
@@ -1,6 +1,8 @@
 using Measurements, SpecialFunctions, QuadGK, Calculus
 using Test, LinearAlgebra, Statistics, Unitful, Printf, Aqua
-Aqua.test_all(Measurements)
+
+
+#Aqua.test_all(Measurements)
```
Comment (giordano) [inline test/runtests.jl:5]: Why?

**Review by giordano — COMMENTED** [review]
Thanks for tackling this!

I'm also inclined towards removing the Juno stuff altogether: as far as I know it's broken, I doubt anyone uses it, Juno itself is discontinued. But this is besides the point of this PR.

File: test/runtests.jl (hunk 8 above)
Comment (longemen3000) [reply test/runtests.jl:5]: ohh, sorry, that wasn't meant to be in the PR, but Aqua has problems in Formatting extensions ([#105](https://github.com/JuliaTesting/Aqua.jl/issues/105), ([#107](https://github.com/JuliaTesting/Aqua.jl/issues/107)

**Comment by longemen3000** [comment]
Finally, tests pass hahah. the `__init__` function requires that particular arrangement to work on earlier versions (based on https://github.com/SciML/DiffEqBase.jl/commit/c2de920a88a39814ac368f703c40fbc282a9c078) . i think that there will be some problems with Aqua.jl afterwards, because there isn't any compat of the `@required` functions

File: src/Measurements.jl [hunk 9]
```diff
@@ -85,9 +83,17 @@ end
 # Start from 1, 0 is reserved to derived quantities
 const tag_counter = Threads.Atomic{UInt64}(1)
 
+
+if !isdefined(Base, :get_extension)
+    using Requires
+end
```
Comment (giordano) [inline src/Measurements.jl:89]: There's a trailing whitespace
```suggestion
end
```

File: src/Measurements.jl [hunk 10]
```diff
@@ -85,9 +83,17 @@ end
...
+
 function __init__()
-    @require Unitful="1986cc42-f94f-5a68-af5c-568840ba703d" include("unitful.jl")
-    @require SpecialFunctions="276daf66-3868-5448-9aa4-cd146d93841b" include("special-functions.jl")
+    @static if !isdefined(Base, :get_extension)
+        @require Unitful="1986cc42-f94f-5a68-af5c-568840ba703d" include("../ext/MeasurementsUnitfulExt.jl")
+        @require SpecialFunctions="276daf66-3868-5448-9aa4-cd146d93841b" include("../ext/MeasurementsSpecialFunctionsExt.jl")
+        @require Juno="e5e0dc1b-0480-54bc-9374-aad01c23163d" include("../ext/MeasurementsJunoExt.jl")
+    end
 end
```
Comment (giordano) [inline src/Measurements.jl:97]: At this point just use
```julia
@static if !isdefined(Base, :get_extension)
    using Requires
    function __init__()
        @require Unitful="1986cc42-f94f-5a68-af5c-568840ba703d" include("../ext/MeasurementsUnitfulExt.jl")
        @require SpecialFunctions="276daf66-3868-5448-9aa4-cd146d93841b" include("../ext/MeasurementsSpecialFunctionsExt.jl")
        @require Juno="e5e0dc1b-0480-54bc-9374-aad01c23163d" include("../ext/MeasurementsJunoExt.jl")
    end
end
```
Having a duplicate `if` looks pointless and you avoid having an `__init__` function completely when not needed.

File: test/runtests.jl (hunk 8 above)
Comment (giordano) [reply test/runtests.jl:5]: Doesn't look like https://github.com/PainterQubits/Unitful.jl/pull/622 had to remove this.  Can it be brought back?  It's pointless to have Aqua has a test dependency and don't use it (and I hate commented out code: code is either used or it doesn't exist)

File: Project.toml [hunk 11]
```diff
@@ -11,6 +11,16 @@ RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
...
+Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
+
+[extensions]
+MeasurementsJunoExt = "Juno"
+MeasurementsSpecialFunctionsExt = "SpecialFunctions"
+MeasurementsUnitfulExt = "Unitful"
+
 [compat]
 Calculus = "0.4.1, 0.5"
 RecipesBase = "0.6.0, 0.7, 0.8, 1.0"
```
Comment (giordano) [inline Project.toml:26]: Also [`src/plot-recipes.jl`](https://github.com/JuliaPhysics/Measurements.jl/blob/f1d6e89b4ef6f24f5cb710c4f71d35afabf76af0/src/plot-recipes.jl) should be an extension.

File: ext/MeasurementsRecipesBaseExt.jl [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    xerror := uncertainty.(x)
+    value.(x), y
+end
+
+@recipe function f(x::AbstractArray, y::AbstractArray{<:Measurement})
+    yerror := uncertainty.(y)
+    x, value.(y)
+end
+
+end
```
Comment (giordano) [inline ext/MeasurementsRecipesBaseExt.jl:39]:
```suggestion
end

```

File: test/runtests.jl [hunk 13]
```diff
@@ -1,5 +1,7 @@
 using Measurements, SpecialFunctions, QuadGK, Calculus
 using Test, LinearAlgebra, Statistics, Unitful, Printf, Aqua
+
+
```
Comment (giordano) [inline test/runtests.jl:4]:
```suggestion
```

File: ext/MeasurementsRecipesBaseExt.jl
Comment (giordano) [inline ext/MeasurementsRecipesBaseExt.jl:1]: Shouldn't [`src/plot-recipes.jl`](https://github.com/JuliaPhysics/Measurements.jl/blob/f1d6e89b4ef6f24f5cb710c4f71d35afabf76af0/src/plot-recipes.jl)  be deleted (in practice renamed to this file)?

File: ext/MeasurementsSpecialFunctionsExt.jl
Comment (giordano) [inline ext/MeasurementsSpecialFunctionsExt.jl:1]: Shouldn't [`src/special-functions.jl`](https://github.com/JuliaPhysics/Measurements.jl/blob/f1d6e89b4ef6f24f5cb710c4f71d35afabf76af0/src/special-functions.jl) be deleted?

File: ext/MeasurementsRecipesBaseExt.jl
Comment (longemen3000) [reply ext/MeasurementsRecipesBaseExt.jl:1]: yes

File: ext/MeasurementsRecipesBaseExt.jl [hunk 14]
```diff
@@ -44,3 +35,5 @@ end
     yerror := uncertainty.(y)
     x, value.(y)
 end
+
+end
```
Comment (giordano) [inline ext/MeasurementsRecipesBaseExt.jl:39]:
```suggestion
end

```

File: Project.toml [hunk 15]
```diff
@@ -11,6 +11,18 @@ RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
...
+Juno = "e5e0dc1b-0480-54bc-9374-aad01c23163d"
+RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
+SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
+Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
+
+[extensions]
+MeasurementsJunoExt = "Juno"
+MeasurementsSpecialFunctionsExt = "SpecialFunctions"
+MeasurementsUnitfulExt = "Unitful"
+MeasurementsRecipesBaseExt = "Unitful"
```
Comment (giordano) [inline Project.toml:24]:
```suggestion
MeasurementsRecipesBaseExt = "RecipesBase"
```
?

Comment (longemen3000) [reply Project.toml:24]: that is the error i was trying to find hahah

**Comment by longemen3000** [comment]
the loading of RecipesBase was done in a backwards-compatible (https://pkgdocs.julialang.org/dev/creating-packages/#Transition-from-normal-dependency-to-extension) but aqua complains about that (and is the reason the Aqua test fail)

File: test/runtests.jl [hunk 16]
```diff
@@ -1,6 +1,12 @@
 using Measurements, SpecialFunctions, QuadGK, Calculus
 using Test, LinearAlgebra, Statistics, Unitful, Printf, Aqua
-Aqua.test_all(Measurements)
+
+if !isdefined(Base,:get_extension)
+    Aqua.test_all(Measurements)
+else
+    Aqua.test_all(Measurements,test_stale_deps = false)
```
Comment (giordano) [inline test/runtests.jl]:
```suggestion
    Aqua.test_all(Measurements; stale_deps=false)
```

**Comment by longemen3000** [comment]
the tests are now explicitly failing for https://github.com/JuliaTesting/Aqua.jl/issues/105 (the Pkg order is different than the order Aqua expects), should be fixed by https://github.com/JuliaTesting/Aqua.jl/pull/106 ?

**Comment by giordano** [comment]
Can't you just change the order to make the test pass or am I missing something?

**Comment by longemen3000** [comment]
i will try, hope that a `Pkg.resolve()` does not change the proposed order

**Comment by longemen3000** [comment]
Tests seems to pass (now with Aqua.jl)

File: test/runtests.jl [hunk 17]
```diff
@@ -1,6 +1,12 @@
 using Measurements, SpecialFunctions, QuadGK, Calculus
 using Test, LinearAlgebra, Statistics, Unitful, Printf, Aqua
-Aqua.test_all(Measurements)
+
+if !isdefined(Base,:get_extension)
+    Aqua.test_all(Measurements)
+else
+    Aqua.test_all(Measurements,stale_deps = false)
+    Aqua.test_stale_deps(Measurements,ignore = [:RecipesBase,:Requires])
```
Comment (giordano) [inline test/runtests.jl:8]: Style:
```suggestion
    Aqua.test_all(Measurements; stale_deps=false)
    Aqua.test_stale_deps(Measurements; ignore=[:RecipesBase, :Requires])
```
I'm strongly of the opinion that everybody should ***always*** start keyword arguments with `;` if they don't want to fall into traps.

**Comment by giordano** [comment]
Ok, I think this is in a good shape now. Thanks again for this!

File: src/Measurements.jl [hunk 18]
```diff
@@ -128,6 +121,19 @@ include("math.jl")
 include("linear_algebra.jl")
 include("show.jl")
 include("parsing.jl")
-include("plot-recipes.jl")
+
+if !isdefined(Base,:get_extension)
+    using Requires
+    using RecipesBase
+    include("../ext/MeasurementsRecipesBaseExt.jl")
```
Comment (LebedevRI) [inline src/Measurements.jl:128]: Is it intentional that only this single extension is outside of the next block?

Comment (longemen3000) [reply src/Measurements.jl:128]: this package supports julia 1.0. for some reason, using one `isdefined` block failed.

Comment (LebedevRI) [reply src/Measurements.jl:128]: I see, thanks.

**Pull request merged**